core/src/main/scala/tech/beshu/ror/accesscontrol/domain/userAndGroups.scala

```diff
@@ -138,32 +147,59 @@ object PermittedGroupIds {
     }
 
     def matches(groupId: GroupId): Boolean = {
-      permittedGroupIds.matcher.`match`(groupId)
+      groupIds.matcher.`match`(groupId)
     }
   }
 }
 
-final case class ResolvablePermittedGroupIds(permittedGroupIds: UniqueNonEmptyList[RuntimeMultiResolvableVariable[GroupIdLike]]) {
-  def resolveGroupIds[B <: BlockContext](blockContext: B): Option[PermittedGroupIds] = {
+final case class ResolvableGroupIds(groupIds: UniqueNonEmptyList[RuntimeMultiResolvableVariable[GroupIdLike]]) {
+  def resolveGroupIds[B <: BlockContext](blockContext: B): Option[GroupIds] = {
     UniqueNonEmptyList
-      .from(resolveAll(permittedGroupIds.toNonEmptyList, blockContext))
-      .map(PermittedGroupIds.apply)
+      .from(resolveAll(groupIds.toNonEmptyList, blockContext))
+      .map(GroupIds.apply)
   }
 }
 
-sealed trait GroupsLogic {
-  val permittedGroupIds: PermittedGroupIds
-}
+sealed trait GroupsLogic
 
 object GroupsLogic {
-  final case class Or(override val permittedGroupIds: PermittedGroupIds) extends GroupsLogic
-  final case class And(override val permittedGroupIds: PermittedGroupIds) extends GroupsLogic
+
+  sealed trait PositiveGroupsLogic extends GroupsLogic {
+    val permittedGroupIds: GroupIds
+  }
+
+  final case class Or(override val permittedGroupIds: GroupIds) extends PositiveGroupsLogic
+
+  final case class And(override val permittedGroupIds: GroupIds) extends PositiveGroupsLogic
+
+  sealed trait NegativeGroupsLogic extends GroupsLogic {
+    val forbiddenGroupIds: GroupIds
+  }
+
+  final case class NotAnyOf(override val forbiddenGroupIds: GroupIds) extends NegativeGroupsLogic
+
+  final case class NotAllOf(override val forbiddenGroupIds: GroupIds) extends NegativeGroupsLogic
+
+  sealed trait CombinedGroupsLogic extends GroupsLogic {
```

Comment:
coutoPL comment migrated from old PR:

**_WDYT about defining it like this:

final case class CombinedGroupsLogic(permittedGroupsLogic: PositiveGroupsLogic, forbiddenGroupsLogic: NegativeGroupsLogic)
Can it be done like that?_**

Reply:
Done, it simplified the logic and allowed to combine nagative rules also with AND